### PR TITLE
Fix case of empty string in String.lastIndexOf and String.indexOf TG-592

### PR DESCRIPTION
**Pull request by romainbrenguier**
The result in that case is expected to be 0 for indexOf and length() for lastIndexOf

### CODE REVIEW

File: regression/jbmc-strings/StringLastIndexOf/Test.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+public class Test {
+    public boolean check(String input_String1, String input_String2) {
+        if(input_String1 != null && input_String2 != null) {
+            int lio = input_String1.lastIndexOf(input_String2);
```
Comment (allredj) [inline regression/jbmc-strings/StringLastIndexOf/Test.java]: We also need a test for the `lastIndexOf(int ch, int fromIndex)` method.

**Review by allredj — APPROVED** [review]
LGTM, but it would be good to also have a test for `lastIndexOf(int ch, int fromIndex)`.

**Comment by allredj** [comment]
Some tests are failing. Could it be due to having changed the value of `from_index`?

**Comment by romainbrenguier** [comment]
@allredj 
> Some tests are failing. Could it be due to having changed the value of from_index?

The test for `lastIndexOf(C)` was failing because we were not checking that the result was smaller than the length of the string, I have now added that to the axioms.

File: src/solvers/refinement/string_constraint_generator_indexof.cpp [hunk 2]
```diff
@@ -238,6 +247,11 @@ exprt string_constraint_generatort::add_axioms_for_last_index_of_string(
     needle);
   axioms.push_back(a5);
 
+  const implies_exprt a6(
+      equal_exprt(needle.length(), from_integer(0, index_type)),
+      equal_exprt(offset, from_index));
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_indexof.cpp]: bad indent

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
Getting this kind of axiomatisation right is challenging so I can believe that this is a fix.  I don't know the current state of the algorithm well enough to comment on the performance impact but I'll trust you've looked at that.  Will approve and / or merge once CI has completed.

**Comment by romainbrenguier** [comment]
@martin-cs this will only affect performance for functions using indexOf and lastIndexOf and they need to be fix anyway, the fix for lastIndexOf(char) should actually performances (from the tests I've run)

@allredj good idea to add that to the test, made me find a problem with lastIndexOf(char). Can you have a look at the commit that fixes that (81c5fd4) ?

**Review by allredj — APPROVED** [review]
LGTM.

**Pull request merged**